### PR TITLE
Rdcc 453

### PR DESCRIPTION
**Pull request by KOTLAPRA23**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/RDCC-452

### Change description ###

delete roles for user

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ X] No
```

### CODE REVIEW

File: src/functionalTest/java/uk/gov/hmcts/reform/userprofileapi/DeleteRolesToExistingUserFuncTest.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+
+        GetUserProfileWithRolesResponse resourceForDeleteCheck =
+                testRequestHandler.sendGet(
+                        "/v1/userprofile/" + resource.getIdamId() + "/roles",
+                        GetUserProfileWithRolesResponse.class
+                );
+        LOG.info("Roles addroles call" + resource2);
+        assertThat(resourceForDeleteCheck.getRoles().size()).isNotNull();
+        assertThat(resourceForDeleteCheck.getRoles().size()).isEqualTo(2);
```
Comment (shreedhar-hmcts) [inline src/functionalTest/java/uk/gov/hmcts/reform/userprofileapi/DeleteRolesToExistingUserFuncTest.java]: Better to check if exact roles are deleted from SIDAM, just check in GetUserProfileWithRolesResponse if roles deleted are vanished from SIDAM and the roles which were not deleted are still intact in SIDAM.

File: src/functionalTest/java/uk/gov/hmcts/reform/userprofileapi/AddRolesToExistingUserFuncTest.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        LOG.info("after addroles call" + resource1);
+
+        GetUserProfileWithRolesResponse resource2 =
+                testRequestHandler.sendGet(
+                        "/v1/userprofile/" + resource.getIdamId() + "/roles",
+                        GetUserProfileWithRolesResponse.class
+                );
+        LOG.info("Roles addroles call" + resource2);
+        assertThat(resource2.getRoles().size()).isNotNull();
+        assertThat(resource2.getRoles().size()).isEqualTo(3);
```
Comment (shreedhar-hmcts) [inline src/functionalTest/java/uk/gov/hmcts/reform/userprofileapi/AddRolesToExistingUserFuncTest.java]: Along with size it is also better to check exact roles which got added newly from GetUserProfileWithRolesResponse

File: Jenkinsfile_CNP [hunk 3]
```diff
@@ -32,7 +32,7 @@ withPipeline(type, product, component) {
 
     env.S2S_URL_TO_TEST = "http://rpe-service-auth-provider-aat.service.core-compute-aat.internal"
     env.IDAM_URL_TO_TEST = "https://idam-api.aat.platform.hmcts.net"
-    env.TEST_URL = "http://rd-user-profile-api-aat.service.core-compute-aat.internal"
+    env.TEST_URL = "http://rd-user-profile-api-preview.service.core-compute-preview.internal"
```
Comment (shreedhar-hmcts) [inline Jenkinsfile_CNP]: I think better keep aat here

File: src/main/java/uk/gov/hmcts/reform/userprofileapi/controller/UserProfileController.java [hunk 4]
```diff
@@ -25,16 +27,7 @@
...
-import uk.gov.hmcts.reform.userprofileapi.client.CreateUserProfileResponse;
-import uk.gov.hmcts.reform.userprofileapi.client.GetUserProfileResponse;
-import uk.gov.hmcts.reform.userprofileapi.client.GetUserProfileWithRolesResponse;
-import uk.gov.hmcts.reform.userprofileapi.client.GetUserProfilesRequest;
-import uk.gov.hmcts.reform.userprofileapi.client.GetUserProfilesResponse;
-import uk.gov.hmcts.reform.userprofileapi.client.IdentifierName;
-import uk.gov.hmcts.reform.userprofileapi.client.RequestData;
-import uk.gov.hmcts.reform.userprofileapi.client.UpdateUserProfileData;
-import uk.gov.hmcts.reform.userprofileapi.client.UserProfileIdentifier;
+import uk.gov.hmcts.reform.userprofileapi.client.*;
```
Comment (shreedhar-hmcts) [inline src/main/java/uk/gov/hmcts/reform/userprofileapi/controller/UserProfileController.java]: better to remove *

File: src/main/java/uk/gov/hmcts/reform/userprofileapi/controller/UserProfileController.java [hunk 5]
```diff
@@ -265,12 +256,23 @@
...
             produces = APPLICATION_JSON_UTF8_VALUE
     )
+
     @ResponseBody
-    public ResponseEntity<CreateUserProfileResponse> updateUserProfile(@Valid @RequestBody UpdateUserProfileData updateUserProfileData, @PathVariable String userId) {
+    public ResponseEntity<UserProfileRolesResponse> updateUserProfile(@Valid @RequestBody UpdateUserProfileData updateUserProfileData, @PathVariable String userId
+    ) {
         log.info("Updating user profile");
+        UserProfileRolesResponse userProfileResponse = new UserProfileRolesResponse();
+        if (StringUtils.isEmpty(updateUserProfileData.getRolesAdd())
```
Comment (shreedhar-hmcts) [inline src/main/java/uk/gov/hmcts/reform/userprofileapi/controller/UserProfileController.java]: I am not sure but does StringUtils also check elements in Set as empty ? Can we use CollectionUtils.isEmpty ?

File: src/main/java/uk/gov/hmcts/reform/userprofileapi/service/UserProfileUpdator.java [hunk 6]
```diff
@@ -28,42 +41,116 @@
...
+    @Override
+    public UserProfileRolesResponse updateRoles(UpdateUserProfileData profileData, String userId) {
+        UserProfile userProfile = null;
+        HttpStatus httpStatus = null;
+        UserProfileRolesResponse userProfileRolesResponse =  new UserProfileRolesResponse();
+        userProfile = validateUserStatus(userId);
+        if (!CollectionUtils.isEmpty(profileData.getRolesAdd())) {
+            log.info("Add idam roles for userId :" + userId);
+            AddRoleResponse addRoleResponse = new AddRoleResponse();
+            try (Response response = idamClient.addUserRoles(profileData.getRolesAdd(), userId)) {
```
Comment (shreedhar-hmcts) [inline src/main/java/uk/gov/hmcts/reform/userprofileapi/service/UserProfileUpdator.java]: idamClient.addUserRoles can be reused directly from SidamService which is already created. If in future if addUserRoles functionality of SIDAM is required, then we can reuse from sidamService.

File: src/main/java/uk/gov/hmcts/reform/userprofileapi/util/UserProfileValidator.java [hunk 7]
```diff
@@ -130,4 +129,17 @@ static void validateUserIds(GetUserProfilesRequest getUserProfilesRequest) {
...
         }
     }
+
+    static void validateUserProfileDataAndUserId(UpdateUserProfileData userProfileData, String userId) {
+
+        if (null == userProfileData) {
+
+            throw new RequiredFieldMissingException("No Request Body in the request");
+        } else if (StringUtils.isBlank(userId)
+                || (!CollectionUtils.isEmpty(userProfileData.getRolesAdd()) && userProfileData.getRolesAdd().size() == 0)
```
Comment (shreedhar-hmcts) [inline src/main/java/uk/gov/hmcts/reform/userprofileapi/util/UserProfileValidator.java]: CollectionUtils.isEmpty internally checks null and size both, so I doubt size checking is required

File: src/functionalTest/java/uk/gov/hmcts/reform/userprofileapi/DeleteRolesToExistingUserFuncTest.java (hunk 1 above)
Comment (KOTLAPRA23) [reply src/functionalTest/java/uk/gov/hmcts/reform/userprofileapi/DeleteRolesToExistingUserFuncTest.java]: Yes. Included for roles check as well.

File: src/functionalTest/java/uk/gov/hmcts/reform/userprofileapi/AddRolesToExistingUserFuncTest.java (hunk 2 above)
Comment (KOTLAPRA23) [reply src/functionalTest/java/uk/gov/hmcts/reform/userprofileapi/AddRolesToExistingUserFuncTest.java]: done

File: Jenkinsfile_CNP (hunk 3 above)
Comment (KOTLAPRA23) [reply Jenkinsfile_CNP]: done

File: src/main/java/uk/gov/hmcts/reform/userprofileapi/controller/UserProfileController.java (hunk 4 above)
Comment (KOTLAPRA23) [reply src/main/java/uk/gov/hmcts/reform/userprofileapi/controller/UserProfileController.java]: removed * and added required files in the import.

File: src/main/java/uk/gov/hmcts/reform/userprofileapi/controller/UserProfileController.java (hunk 5 above)
Comment (KOTLAPRA23) [reply src/main/java/uk/gov/hmcts/reform/userprofileapi/controller/UserProfileController.java]: removed the StringUtils and added collectionUtils.isEmpty check

File: src/main/java/uk/gov/hmcts/reform/userprofileapi/util/UserProfileValidator.java (hunk 7 above)
Comment (ConorH994) [reply src/main/java/uk/gov/hmcts/reform/userprofileapi/util/UserProfileValidator.java]: it checks only empty not the size, tests fail when we don't include the size check

Comment (shreedhar-hmcts) [reply src/main/java/uk/gov/hmcts/reform/userprofileapi/util/UserProfileValidator.java]: ok. fine.

**Comment by pcthottempudi** [comment]
Approved

**Pull request merged**